### PR TITLE
Replaced PushStage based CheckContentLength with GraphStage #19834

### PR DESCRIPTION
**Pull request by vans239**

### CODE REVIEW

**Comment by johanandren** [comment]
OK TO TEST

File: akka-http-core/src/main/scala/akka/http/impl/engine/rendering/RenderSupport.scala [hunk 1]
```diff
@@ -69,28 +70,45 @@ private object RenderSupport {
...
-      if (sent < length)
-        throw InvalidContentLengthException(s"HTTP message had declared Content-Length $length but entity data stream amounts to ${length - sent} bytes less")
-      ctx.finish()
-    }
+        private var sent = 0L
+
+        override def onPush(): Unit = {
+          val elem = grab(in)
+          sent += elem.length
+          if (sent > length) {
```
Comment (viktorklang) [inline akka-http-core/src/main/scala/akka/http/impl/engine/rendering/RenderSupport.scala]: put the most likely to succeed branch first.

**Comment by patriknw** [comment]
LGTM

File: akka-http-core/src/main/scala/akka/http/impl/engine/rendering/RenderSupport.scala [hunk 2]
```diff
@@ -69,28 +70,45 @@ private object RenderSupport {
...
 
   object CheckContentLengthTransformer {
-    def flow(contentLength: Long) = Flow[ByteString].transform(() ⇒
-      new CheckContentLengthTransformer(contentLength)).named("checkContentLength")
+    def flow(contentLength: Long) = Flow[ByteString].via(new CheckContentLengthTransformer(contentLength))
   }
 
-  class CheckContentLengthTransformer(length: Long) extends PushStage[ByteString, ByteString] {
-    var sent = 0L
+  class CheckContentLengthTransformer(length: Long) extends SimpleLinearGraphStage[ByteString] {
```
Comment (johanandren) [inline akka-http-core/src/main/scala/akka/http/impl/engine/rendering/RenderSupport.scala]: final?

File: akka-http-core/src/main/scala/akka/http/impl/engine/rendering/RenderSupport.scala (hunk 1 above)
Comment (vans239) [inline akka-http-core/src/main/scala/akka/http/impl/engine/rendering/RenderSupport.scala]: Fixed. But does it make any sense. Jit will optimize this

**Comment by vans239** [comment]
Changed according to feedback.

**Comment by vans239** [comment]
Error in build:
`java.lang.RuntimeException: akka-stream: Binary compatibility check failed!`
Looks like it isn't related to commit

**Comment by agolubev** [comment]
@vans239 you can add FilterAnyProblemStartingWith("akka.http.impl") to exclude all internal API classes. Perhaps it's been done already and you'll see as rebase to master head

**Comment by vans239** [comment]
@agolubev Thanks for help. Added mima exclude for akka.http.impl

**Comment by patriknw** [comment]
looks ready to me, any more reviewers?

**Comment by drewhk** [comment]
LGTM

**Comment by drewhk** [comment]
refs #19834

**Pull request merged**